torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -377,6 +378,318 @@ bool IsBlockReturnTypeSame(Node* n) {
   return true;
 }
 
+c10::optional<at::Tensor> ComputeConstantFolding(Node* n, int opset_version) {
+  if (n->inputs().size() == 0) {
+    return c10::nullopt;
+  }
+  std::vector<at::Tensor> inputTensorValues;
+  for (auto i = 0; i < n->inputs().size(); i++) {
+    if (TensorTypePtr input_type = n->input(i)->type()->cast<TensorType>()) {
+      if (!ConstantValueMap::HasValue(n->input(i)->debugName())) {
+        return c10::nullopt;
+      }
+      auto tensor_value = ConstantValueMap::GetValue(n->input(i)->debugName());
+      inputTensorValues.emplace_back(tensor_value);
+    }
+  }
+  if (inputTensorValues.size() < n->inputs().size()) {
+    return c10::nullopt;
+  }
+  try {
+    return onnx_constant_fold::runTorchBackendForOnnx(
+        n, inputTensorValues, opset_version);
+  } catch (const std::exception& ex) {
+    TORCH_WARN(
+        "Constant folding in symbolic shape inference fails: ", ex.what());
+    return c10::nullopt;
+  }
+}
+
+// When the Reshape node's two inputs are constant, compute the output shape.
+// The reshape value 0 and -1 are converted to the real value explicitly.
+std::vector<int64_t> ComputeShapeFromReshape(
+    const std::vector<int64_t>& input_shape,
+    const std::vector<int64_t>& reshape) {
+  TORCH_INTERNAL_ASSERT(
+      input_shape.size() > 0 || reshape.size() > 0,
+      "Reshape node should have at least one input size > 0 when constant folding.");
+  // Case: reshape.size() == 0
+  // %22 : int[] = prim::Constant[value=annotate(List[int], [])]()
+  // %15 : Float(requires_grad=0, device=cpu) = aten::view(%1, %22)
+  if (reshape.size() == 0) {
+    return input_shape;
+  }
+  if (input_shape.size() == 0) {
+    return reshape;
+  }
+  auto reshape_has_zero = false;
+  auto reshape_size = static_cast<int>(reshape.size());
+  for (auto i = 0; i < reshape_size; i++) {
+    if (reshape[i] == 0) {
+      reshape_has_zero = true;
+      break;
+    }
+  }
+  int minus_one_pos = -1;
+  auto input_shape_size = static_cast<int>(input_shape.size());
+  for (auto i = 0; i < input_shape_size; i++) {
+    if (reshape[i] == -1) {
+      minus_one_pos = i;
+      break;
+    }
+  }
+  if (!reshape_has_zero && minus_one_pos == -1) {
+    return reshape;
+  }
+  std::vector<int64_t> final_shape;
+  // shape_ratio is used to calculate the real value of -1.
+  // it is updated on each dimension to avoid overflow.
```

Comment:
nice, but it seems unlikely for number of elements in a tensor to exceed 2^63-1...